### PR TITLE
Net Reproductive Rate

### PR DESCRIPTION
**Pull request by krpaulson**
## Describe changes

Add function to calculate net reproductive rate.

## What issues are related

Related to #19 

### Packages Repositories

* [x] Have you read the [contributing guidelines](https://github.com/ihmeuw-demographics/packageTemplate/wiki#guide-to-r-package-development) for `ihmeuw-demographics` R packages?
* [x] Have you successfully run `devtools::check()` locally?
* [x] Have you updated or added function (and vignette if applicable) documentation? Did you update the 'man' and 'NAMESPACE' files with `devtools::document()`?
* [x] Have you added in tests for the changes included in the PR?
* [x] Do the changes follow the `ihmeuw-demographics` [code style](https://github.com/ihmeuw-demographics/packageTemplate/wiki/Code-style-guide)?
* [ ] Do your changes need to be immediately included in a new build of [`docker-base`](https://github.com/ihmeuw-demographics/docker-base) or [`docker-internal`](https://github.com/ihmeuw-demographics/docker-internal)? If so follow directions in those repositories to rebuild and redeploy the images.

## Details of PR

This PR is not yet ready for review. It was started in a group working session but still needs to be debugged and completed.

### CODE REVIEW

File: R/gen_nrr.R [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+#' Preston Demography textbook page 113.
+#'
+#' @examples
+#' dt <- data.table::data.table(
+#'   age_start = seq(15, 45, 5),
+#'   age_end = seq(20, 50, 5),
+#'   sex = "female",
+#'   asfr = c(0.00002, 0.009, 0.1, 0.18, 0.19, 0.11, 0.03),
+#'   srb = 1.057,
+#'   nLx = 4.9
```
Comment (chacalle) [inline R/gen_nrr.R:32]: Is this nLx for the entire reproductive age group? Is it more common this will be age-group specific?

File: R/gen_nrr.R [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  assertable::assert_colnames(
+    dt,
+    colnames = c("asfr", "srb", "nLx", id_cols),
+    only_colnames = F,
+    quiet = T
+  )
+
+  # check female
+  assertthat::assert_that(
+    unique(dt$sex) == "female",
```
Comment (chacalle) [inline R/gen_nrr.R:63]: Should use `identical` instead of `==` in case there are more than one

File: R/gen_nrr.R [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+#' @param reproductive_age_end \[`numeric(1)`\]\cr
+#'   Age in years of end of reproductive age span (default 50 years).
+#' @inheritParams gen_u5_ax
+#'
+#' @return dt \[`data.table()`\]\cr
+#'   Data table with `nrr` by `id_cols` (excluding 'age_start' and 'age_end').
+#'
+#' @details
+#' Calculate proportion female: prop_female = (1 / srb) / (1 + (1 / srb)).
+#' Then, calculate NRR: sum(asfr * prop_female * nLx) over age, by id columns.
```
Comment (chacalle) [inline R/gen_nrr.R:20]: These equations are helpful It might be useful to link to a description of nrr somewhere.

File: R/gen_nrr.R [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+#' )
+#' output <- gen_nrr(
+#'   dt,
+#'   id_cols = c("age_start", "age_end", "sex"),
+#'   reproductive_age_start = 15,
+#'   reproductive_age_end = 50
+#' )
+#'
+#' @export
+gen_nrr <- function(dt, id_cols, reproductive_age_start = 15, reproductive_age_end = 50) {
```
Comment (spencerpease) [inline R/gen_nrr.R:42]: What is our convention regarding `gen_*`? Should this be modifying a data.table in-place?

File: tests/testthat/test-gen_nrr.R [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+testthat::test_that("`gen_nrr` gives expected output", {
+
+  output <- gen_nrr(
+    input,
+    id_cols = c("age_start", "age_end", "sex"),
+    reproductive_age_start = 15,
+    reproductive_age_end = 50
+  )
+  testthat::expect_equivalent(output, expected, tolerance = 0.01)
+})
```
Comment (chacalle) [inline tests/testthat/test-gen_nrr.R:26]: Should eventually add in some basic tests for common errors that should be caught by the assertions.

File: R/gen_nrr.R [hunk 6]
```diff
@@ -0,0 +1,91 @@
+#' @title Generate Net Reproductive Rate (NRR)
+#'
+#' @description Calculate Net Reproductive Rate (NRR) from age-specific
+#'   fertility rate (ASFR) and sex-ratio at birth (SRB) and
```
Comment (chacalle) [inline R/gen_nrr.R:4]: and nLx

File: R/gen_nrr.R (hunk 4 above)
Comment (chacalle) [inline R/gen_nrr.R:42]: The `gen_` functions should be reserved for modifying in place. Should this be `nrr_from_nFx_srb_nLx` to be consistent with our lifetable functions?

**Review by chacalle — CHANGES_REQUESTED** [review]

File: R/calculate_nrr.R [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    only_colnames = F,
+    quiet = T
+  )
+
+  # check female
+  assertthat::assert_that(
+    identical(unique(dt$sex), "female"),
+    msg = "`sex` must be 'female'."
+  )
+
```
Comment (erinamay) [inline R/calculate_nrr.R:69]: Should there also be checks for reasonable values (i.e. non negative?) of SRB/asfr/nLx and non-NAs?

File: R/calculate_nrr.R [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+ # calculate ----------------------------------------------------------------
+
+  # calculate proportion births that are female
+  dt[, prop_female := (1 / srb) / (1 + (1 / srb))]
+
+  # calculate NRR
+  dt <- dt[,
+    list(nrr = sum(asfr * prop_female * nLx)),
```
Comment (erinamay) [inline R/calculate_nrr.R:90]: Should na.rm = TRUE be added, or no?

**Review by erinamay — DISMISSED** [review]

File: R/calculate_nrr.R [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+#'   dt,
+#'   id_cols = c("age_start", "age_end", "sex"),
+#'   reproductive_age_start = 15,
+#'   reproductive_age_end = 50
+#' )
+#'
+#' @export
+calculate_nrr <- function(dt,
+                          id_cols,
+                          reproductive_age_start = 15,
```
Comment (chacalle) [inline R/calculate_nrr.R:44]: Should we make these defaults equal to the entire gbd reproductive age span?

**Review by chacalle — APPROVED** [review]

File: R/calculate_nrr.R (hunk 9 above)
Comment (krpaulson) [reply R/calculate_nrr.R:44]: Is that 10-60?

File: R/calculate_nrr.R (hunk 8 above)
Comment (krpaulson) [reply R/calculate_nrr.R:90]: Not if we're requiring asfr, prop_female, and nLx to be non-NA

File: R/calculate_nrr.R (hunk 9 above)
Comment (chacalle) [reply R/calculate_nrr.R:44]: 10-55

**Pull request merged**